corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/api.js

```diff
@@ -28,14 +28,20 @@ hqDefine("cloudcare/js/formplayer/menus/api", function () {
                 if (!params.preview) {
                     // Make sure the user has access to the app
                     if (!appCollection.find(function (app) {
-                        return app.id === params.appId || app.get('copy_of') === params.copyOf;
+                        if (app.id && app.id === params.appId) {
+                            return true;
+                        }
+                        if (app.get('copy_of') && app.get('copy_of') === params.copyOf) {
+                            return true;
+                        }
                     })) {
                         FormplayerFrontend.trigger(
                             'showError',
-                            gettext('Permission Denied')
+                            gettext('The application could not be found')
                         );
                         FormplayerFrontend.trigger('navigateHome');
                         defer.reject();
+                        return;
```

Comment:
Can this return `defer.promise()` so callers can rely on the return value always being a promise?

Reply:
The promise is already being returned in the outer function. This code is inside a inner function that is handling the result of another event call. Here's the simplified version:


```javascript
queryFormplayer: function (params, route) {
    var defer = $.Deferred();
    
    $.when(FormplayerFrontend.getChannel().request("appselect:apps")).done(function (appCollection) {
      if (!params.preview) {
          if (<can't find app>) {
              // show error
              // navigate home
              defer.reject();
              return;
          }
      }
      // make other requests to FP etc.
   }
   return defer.promise();
}
```

Reply:
ohhh, got it now, thanks